platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractCollectionProperty.java

```diff
@@ -623,6 +625,61 @@ protected Value<? extends C> calculateOwnValue(ValueConsumer consumer) {
         }
     }
 
+    /**
+     * A fixed value collector that decorates {@link ElementsFromCollection} with a side effect.
+     */
+    private static class FixedValueCollector<T, C extends Collection<T>> implements Collector<T> {
+        @Nullable
+        private final SideEffect<? super C> sideEffect;
+        private final C value;
+        private final ElementsFromCollection<T> delegate;
+
+        private FixedValueCollector(C value, @Nullable SideEffect<? super C> sideEffect) {
+            this.value = value;
+            this.sideEffect = sideEffect;
+            this.delegate = new ElementsFromCollection<>(value);
+        }
+
+        @Override
+        public Value<Void> collectEntries(ValueConsumer consumer, ValueCollector<T> collector, ImmutableCollection.Builder<T> dest) {
+            Value<Void> result = delegate.collectEntries(consumer, collector, dest);
+            if (sideEffect == null) {
+                return result;
+            }
+            return result.withSideEffect((SideEffect<Void>) unused -> sideEffect.execute(value));
```

Comment:
💅 If you decide to keep it, an idiomatic way seem to be:
```suggestion
            return result.withSideEffect(SideEffect.fixed(value, sideEffect));
```